### PR TITLE
BigDL module serialization based on protobuf

### PR DESCRIPTION
**Pull request by wzhongyuan**
## What changes were proposed in this pull request?

Module Serialization based on protobuf

## How was this patch tested?

unit test, also tested with some common models

## Related links or issues (optional)
https://github.com/intel-analytics/BigDL/issues/608

### CODE REVIEW

**Comment by wzhongyuan** [comment]
To summarize what's been done in this PR
1) Support serialization/deserialization for BigDL module with protobuf 
2) Support 12 data types conversion, other data conversion can be derived from these data types
3)  Default serialization support by reflection
    3.1)  Default one for single module 
    3.2)  Container support for Sequential - like containers
    3.3) Cell serializer for RNN Cell modules
4) All module companion objects are derived from ModuleSerializable or it's derived traits, 13 of which overrides serialization methods
5) For some module classes, just a few ones, they are written with specialized types, I removed them to keep consistent with most modules, one reason for this is to avoid having multiple class objects for same modules (generic one, float one and double one), in the default implementation, we should have one unique class obj to retrieve the module type , and vice vesa

**Comment by wzhongyuan** [comment]
@jason-dai  @yiheng   Please help review if you have time.  this PR doesn't include version control and python support, I will be working on a separate one for these two. 
Thanks.

**Comment by wzhongyuan** [comment]
PR passed http://10.239.47.211:18888/job/BigDL-PR-Validation/1433/

**Comment by wzhongyuan** [comment]
@yiheng @zhichao-li    Though changes were made to a lot of existing files, most of them are just simple ones to add extend trait in companion objects,  you can specifically focus on the serializer package for code review where the core implementation has been done

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/DataConverter.scala [hunk 1]
```diff
@@ -0,0 +1,620 @@
...
+
+    override def setAttributeValue[T: ClassTag]
+      (attributeBuilder: AttrValue.Builder, value: Any,
+       valueType : universe.Type = null)
+      (implicit ev: TensorNumeric[T]): Unit = {
+      attributeBuilder.setDataType(DataType.TENSOR)
+      if (value != null) {
+        val tensor = value.asInstanceOf[Tensor[T]]
+        val tensorBuilder = BigDLTensor.newBuilder
+        tensor.storage().array().foreach(data => tensorBuilder.addData(ev.toType[Double](data)))
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/DataConverter.scala]: I'm afraid the performance. Should we use byte array to store the tensor conten, like tensorflow

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/DataConverter.scala [hunk 2]
```diff
@@ -0,0 +1,620 @@
...
+/**
+ * DataConverter for [[com.intel.analytics.bigdl.nn.VariableFormat]]
+ */
+  object VariableFormatConverter extends DataConverter {
+
+    override def getAttributeValue[T: ClassTag](attribute: AttrValue)
+      (implicit ev: TensorNumeric[T]): AnyRef = {
+      val format = attribute.getVariableFormatValue
+      format match {
+        case ValFormat.DEFAULT => VariableFormat.Default
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/DataConverter.scala]: ValFormat -> VarFormat? or ParamFormat?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/DataConverter.scala [hunk 3]
```diff
@@ -0,0 +1,620 @@
...
+          serialize(ModuleData(module, Seq[String](), Seq[String]()))
+        attributeBuilder.setBigDLModuleValue(serializableModule)
+      }
+    }
+  }
+
+/**
+ * DataConverter for name list
+ */
+  object NameListConverter extends DataConverter {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/DataConverter.scala]: what is name list? Attributes?

File: spark/dl/src/main/resources/serialization/model.proto [hunk 4]
```diff
@@ -0,0 +1,106 @@
+syntax = "proto3";
```
Comment (wzhongyuan) [inline spark/dl/src/main/resources/serialization/model.proto]: Done

File: spark/dl/src/main/resources/serialization/model.proto [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+package serialization;
+message BigDLModule
+{
+    string name = 1; //model name
```
Comment (wzhongyuan) [inline spark/dl/src/main/resources/serialization/model.proto]: Done

File: spark/dl/src/main/resources/serialization/model.proto [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    repeated BigDLModule subModules = 2; // sub modules
+    BigDLTensor weight = 3; // double weight for each layer
+    BigDLTensor bias = 4; // double bias for each layer
+    repeated string preModules = 5; //pre modules of the same hierarchy
+    repeated string nextModules = 6; //next modules of the same hierachy
+    string moduleType = 7; //module type
+    map<string, AttrValue> attr = 8; // module attribute map
```
Comment (wzhongyuan) [inline spark/dl/src/main/resources/serialization/model.proto]: Done

File: spark/dl/src/main/resources/serialization/model.proto [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+}
+enum ValFormat {
```
Comment (wzhongyuan) [inline spark/dl/src/main/resources/serialization/model.proto]: Rename to VarFormat

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/DataConverter.scala [hunk 8]
```diff
@@ -0,0 +1,620 @@
...
+                                      valueType : universe.Type = null)
+    (implicit ev: TensorNumeric[T]) : Unit
+}
+
+/**
+ * General implementation of [[DataConverter]], it provides the conversion entry for all types
+ */
+object DataConverter extends DataConverter{
+
+  val typePlaceHolder = universe.typeOf[DataConverter]
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/DataConverter.scala]: Done

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/DataConverter.scala [hunk 9]
```diff
@@ -0,0 +1,620 @@
...
+    implicit ev: TensorNumeric[T]) : AnyRef
+
+/**
+ * Set attribute value to protobuf format
+ * @tparam T data type
+ * @param attributeBuilder  the attribute value writable instance
+ * @param value the value to be written to protobuf file
+ * @param valueType the type of the value to help set the data type
+ */
+  def setAttributeValue[T : ClassTag](attributeBuilder : AttrValue.Builder, value: Any,
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/DataConverter.scala]: it' used in sub class (or specifically in sub object which use the TensorNumeric to infer the type or transfer data)

**Comment by wzhongyuan** [comment]
@yiheng  As discussed offline , I've made the refinement as
1) Rename to bigdl for the proto file and add more readable comments
2) Rename ValFormat -> VarFormat
3) DataConverter refinement

File: spark/dl/src/main/resources/serialization/bigdl.proto [hunk 10]
```diff
@@ -0,0 +1,110 @@
+syntax = "proto3";
+package serialization;
+message BigDLModule
+{
+    string name = 1; //module name
+    repeated BigDLModule subModules = 2; // sub modules
+    BigDLTensor weight = 3; // weight for each layer, serialized data are stored as either float or double
+    BigDLTensor bias = 4; // bias for each layer
+    repeated string preModules = 5; //pre modules of the same hierarchy
```
Comment (yiheng) [inline spark/dl/src/main/resources/serialization/bigdl.proto]: there's an assumption that the layer name is unique. But currently, we can't guarantee this in code level. Do we check this before we serialize the model?

File: spark/dl/src/main/resources/serialization/bigdl.proto [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    repeated string nextModules = 6; //next modules of the same hierachy
```
Comment (yiheng) [inline spark/dl/src/main/resources/serialization/bigdl.proto]: there's an assumption that the layer name is unique. But currently, we can't guarantee this in code level. Do we check this before we serialize the model?

File: spark/dl/src/main/resources/serialization/bigdl.proto [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+    string moduleType = 7; //module type to identify a module like linear, graph, etc
```
Comment (yiheng) [inline spark/dl/src/main/resources/serialization/bigdl.proto]: can the moduleType be an enum?

Comment (yiheng) [inline spark/dl/src/main/resources/serialization/bigdl.proto]: what if two module has same name?

File: spark/dl/src/main/resources/serialization/bigdl.proto [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    map<string, AttrValue> attr = 8; // module attributes
+}
+enum VarFormat {
```
Comment (yiheng) [inline spark/dl/src/main/resources/serialization/bigdl.proto]: how 3rd project extend this?

File: spark/dl/src/main/resources/serialization/bigdl.proto [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+    ONE_D = 2;
+    IN_OUT = 3;
+    OUT_IN = 4;
+    IN_OUT_KW_KH = 5;
+    OUT_IN_KW_KH = 6;
+    GP_OUT_IN_KW_KH = 7;
+    GP_IN_OUT_KW_KH = 8;
+    OUT_IN_KT_KH_KW = 9;
+}
+enum InitMethodType {
```
Comment (yiheng) [inline spark/dl/src/main/resources/serialization/bigdl.proto]: how 3rd project extend this?

File: spark/dl/src/main/resources/serialization/bigdl.proto [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+    RANDOM_UNIFORM = 1;
+    RANDOM_UNIFORM_PARAM = 2;
+    RANDOM_NORMAL = 3;
+    ZEROS = 4;
+    ONES = 5;
+    CONST = 6;
+    XAVIER = 7;
+    BILINEARFILLER = 8;
+}
+enum RegularizerType {
```
Comment (yiheng) [inline spark/dl/src/main/resources/serialization/bigdl.proto]: how 3rd project extend this?

File: spark/dl/src/main/resources/serialization/bigdl.proto [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+    L1L2Regularizer = 0;
+    L1Regularizer = 1;
+    L2Regularizer = 2;
+}
+message InitMethod {
+    InitMethodType methodType = 1;
+    repeated double data = 2;
```
Comment (yiheng) [inline spark/dl/src/main/resources/serialization/bigdl.proto]: what if an initMethod need 2 or 3 parameters?

File: spark/dl/src/main/resources/serialization/bigdl.proto [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+}
+message BigDLTensor{
+    DataType datatype = 1;
+    repeated int32 size = 2; //size of tensor
+    repeated float float_data = 3; //tensor element
```
Comment (yiheng) [inline spark/dl/src/main/resources/serialization/bigdl.proto]: bytes?

see
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/tensor.proto#L36

File: spark/dl/src/main/resources/serialization/bigdl.proto (hunk 11 above)
Comment (wzhongyuan) [inline spark/dl/src/main/resources/serialization/bigdl.proto]: Agree it's better to do this sanity check, but there is one exception for MapTable which replicates the module `n` times where n is the input table size. all the modules have the same reference thus threw exceptions, need to think about how to handle this situation...

File: spark/dl/pom.xml [hunk 18]
```diff
@@ -34,7 +34,7 @@
         <dependency>
             <groupId>com.google.protobuf</groupId>
             <artifactId>protobuf-java</artifactId>
-            <version>3.0.0</version>
+            <version>3.2.0</version>
```
Comment (yiheng) [inline spark/dl/pom.xml]: why update the version number？

Comment (wzhongyuan) [inline spark/dl/pom.xml]: Nothing special, just upgrade to recommended version. As discussed offline , I reverted it back to 3.0.0

**Comment by wzhongyuan** [comment]
@yiheng  As discussed offline, I will split the PR into smaller ones

**Comment by yiheng** [comment]
it's conflict. can you rebase it?

**Comment by wzhongyuan** [comment]
I am going to close this one and other following splitted ones will be opened soon for review

**Comment by wzhongyuan** [comment]
Will be committed in splitted PRs, so am closing it

**Pull request closed without merging**